File: push-be/internal/domain/usecase/job.go
package usecase

import (
	"context"
	"encoding/json"
	"errors"
	"strings"
	"time"

	"github.com/google/uuid"

	"push-be/internal/domain"
)

type JobService struct {
	jobs         domain.JobStore
	analyses     domain.GapAnalysisStore
	applications domain.ApplicationStore
	evidence     domain.EvidenceStore
	approvals    *ApprovalService
	ops          domain.OperationStore
	uow          domain.UnitOfWork
	ai           *AIGate
}

type AIGate struct {
	ManagedConfigured bool
	ManagedKey        string
	ByokEnabled       bool
	Keys              domain.AiKeyStore
	Cipher            domain.KeyCipher
	Chat              domain.ChatCompleter
}

func (g *AIGate) Check(ctx context.Context, userID uuid.UUID, ai *domain.AiOptions) *domain.Error {
	return domain.CheckAICredentials(ai, g.ManagedConfigured, func(provider string) bool {
		if !g.ByokEnabled || g.Keys == nil {
			return false
		}
		_, err := g.Keys.Get(ctx, userID, provider)
		return err == nil
	})
}

func (g *AIGate) resolveKey(ctx context.Context, userID uuid.UUID, ai *domain.AiOptions) (string, *domain.Error) {
	if ai.CredentialMode == "MANAGED" {
		return g.ManagedKey, nil
	}
	k, err := g.Keys.Get(ctx, userID, ai.Provider)
	if err != nil {
		return "", domain.IntegrationRequired("no BYOK key configured for " + ai.Provider)
	}
	if g.Cipher == nil {
		return "", domain.NotConfigured("BYOK encryption is not configured")
	}
	key, err := g.Cipher.Decrypt(k.Ciphertext, k.Nonce, userID, ai.Provider)
	if err != nil {
		return "", domain.IntegrationRequired("BYOK key could not be decrypted")
	}
	return key, nil
}

func (g *AIGate) Complete(ctx context.Context, userID uuid.UUID, ai *domain.AiOptions, system, user string) (*domain.AICompletion, *domain.Error) {
	if err := g.Check(ctx, userID, ai); err != nil {
		return nil, err
	}
	if ai.Provider != "OPENAI" || g.Chat == nil {
		return nil, domain.NotConfigured("AI provider " + ai.Provider + " is not supported")
	}
	key, derr := g.resolveKey(ctx, userID, ai)
	if derr != nil {
		return nil, derr
	}
	c, err := g.Chat.Chat(ctx, key, ai.Model, system, user)
	if err != nil {
		return nil, domain.ProviderError("AI provider request failed")
	}
	return c, nil
}

func NewJobService(jobs domain.JobStore, analyses domain.GapAnalysisStore, applications domain.ApplicationStore, evidence domain.EvidenceStore, approvals *ApprovalService, ops domain.OperationStore, uow domain.UnitOfWork, ai *AIGate) *JobService {
	return &JobService{jobs: jobs, analyses: analyses, applications: applications, evidence: evidence, approvals: approvals, ops: ops, uow: uow, ai: ai}
}

type CreateJobInput struct {
	Company      string
	Title        string
	SourceKind   domain.JobSourceKind
	SourceURL    *string
	SourceText   string
	Requirements []string
	Preferred    []string
	Deadline     *time.Time
	Language     string
}

func (s *JobService) List(ctx context.Context, userID uuid.UUID, f domain.JobFilter) (domain.Page[domain.JobPosting], error) {
	limit := f.Page.EffectiveLimit()
	items, err := s.jobs.List(ctx, userID, f, limit+1)
	if err != nil {
		return domain.Page[domain.JobPosting]{}, domain.Internal()
	}
	return domain.NewPage(items, limit, func(j domain.JobPosting) domain.Cursor {
		return domain.Cursor{CreatedAt: j.CreatedAt, ID: j.ID}
	}), nil
}

func (s *JobService) Get(ctx context.Context, userID, id uuid.UUID) (*domain.JobPosting, error) {
	j, err := s.jobs.Get(ctx, userID, id)
	if err != nil {
		if errors.Is(err, domain.ErrNotFound) {
			return nil, domain.NotFound()
		}
		return nil, domain.Internal()
	}
	return j, nil
}

func (s *JobService) Create(ctx context.Context, userID uuid.UUID, in CreateJobInput) (*domain.JobPosting, error) {
	if err := validateJobInput(in.Company, in.Title, in.SourceKind, in.SourceURL, in.SourceText); err != nil {
		return nil, err
	}
	now := time.Now().UTC()
	j := &domain.JobPosting{
		ID: uuid.New(), UserID: userID, Revision: 1,
		Company: in.Company, Title: in.Title, SourceKind: in.SourceKind,
		SourceURL: in.SourceURL, SourceText: in.SourceText,
		Requirements: nonNil(in.Requirements), Preferred: nonNil(in.Preferred),
		Keywords: []string{}, Risks: []string{},
		Deadline: in.Deadline, Language: in.Language,
		CreatedAt: now, UpdatedAt: now,
	}
	if err := s.jobs.Create(ctx, j); err != nil {
		return nil, domain.Internal()
	}
	return j, nil
}

func validateJobInput(company, title string, kind domain.JobSourceKind, sourceURL *string, sourceText string) *domain.Error {
	if len(company) < 1 || len(company) > 200 {
		return domain.ValidationField("company", "company must be 1-200 characters")
	}
	if len(title) < 1 || len(title) > 300 {
		return domain.ValidationField("title", "title must be 1-300 characters")
	}
	if !domain.ValidJobSourceKind(kind) {
		return domain.ValidationField("sourceKind", "must be URL, TEXT or DOM")
	}
	if kind != domain.JobSourceText {
		if sourceURL == nil || *sourceURL == "" {
			return domain.ValidationField("sourceUrl", "required for URL/DOM sources")
		}
	}
	if sourceURL != nil && *sourceURL != "" {
		if err := domain.ValidateHTTPSURL(*sourceURL); err != nil {
			return domain.ValidationField("sourceUrl", "must be an http(s) URL without credentials")
		}
	}
	if n := len(sourceText); n < 1 || n > 200000 {
		return domain.ValidationField("sourceText", "collected source text is required (1-200000 characters)")
	}
	return nil
}

func nonNil(v []string) []string {
	if v == nil {
		return []string{}
	}
	return v
}

type PatchJobInput struct {
	ExpectedRevision int64
	Company          *string
	Title            *string
	Requirements     *[]string
	Preferred        *[]string
	Deadline         *time.Time
	ClearDeadline    bool
}

func (s *JobService) Patch(ctx context.Context, userID, id uuid.UUID, in PatchJobInput) (*domain.JobPosting, error) {
	var out *domain.JobPosting
	err := s.uow.Do(ctx, func(ctx context.Context) error {
		j, err := s.jobs.Get(ctx, userID, id)
		if err != nil {
			if errors.Is(err, domain.ErrNotFound) {
				return domain.NotFound()
			}
			return err
		}
		if in.Company != nil {
			if len(*in.Company) < 1 || len(*in.Company) > 200 {
				return domain.ValidationField("company", "company must be 1-200 characters")
			}
			j.Company = *in.Company
		}
		if in.Title != nil {
			if len(*in.Title) < 1 || len(*in.Title) > 300 {
				return domain.ValidationField("title", "title must be 1-300 characters")
			}
			j.Title = *in.Title
		}
		if in.Requirements != nil {
			j.Requirements = nonNil(*in.Requirements)
		}
		if in.Preferred != nil {
			j.Preferred = nonNil(*in.Preferred)
		}
		if in.ClearDeadline {
			j.Deadline = nil
		} else if in.Deadline != nil {
			j.Deadline = in.Deadline
		}
		j.UpdatedAt = time.Now().UTC()
		if err := s.jobs.Update(ctx, j, in.ExpectedRevision); err != nil {
			return mapRevisionErr(err)
		}
		j.Revision = in.ExpectedRevision + 1
		out = j
		return nil
	})
	if err != nil {
		return nil, err
	}
	return out, nil
}

type AnalyzeInput struct {
	ApplicationID    uuid.UUID
	ExpectedRevision int64
	EvidenceIDs      []uuid.UUID
	AI               *domain.AiOptions
}

func (s *JobService) Analyze(ctx context.Context, userID, jobID uuid.UUID, in AnalyzeInput) (*domain.Operation, error) {
	if err := s.ai.Check(ctx, userID, in.AI); err != nil {
		return nil, err
	}
	var op *domain.Operation
	err := s.uow.Do(ctx, func(ctx context.Context) error {
		j, err := s.jobs.Get(ctx, userID, jobID)
		if err != nil {
			if errors.Is(err, domain.ErrNotFound) {
				return domain.NotFound()
			}
			return err
		}
		if j.Revision != in.ExpectedRevision {
			return domain.RevisionConflict(j.Revision)
		}
		app, err := s.applications.Get(ctx, userID, in.ApplicationID)
		if err != nil {
			if errors.Is(err, domain.ErrNotFound) {
				return domain.NotFound()
			}
			return err
		}
		if app.JobID != jobID {
			return domain.ValidationField("applicationId", "application does not belong to this job")
		}
		pool := []*domain.CareerEvidence{}
		for _, eid := range in.EvidenceIDs {
			e, err := s.evidence.Get(ctx, userID, eid)
			if err != nil {
				return domain.ValidationField("evidenceIds", "evidence "+eid.String()+" not found")
			}
			if e.Archived {
				return domain.ValidationField("evidenceIds", "evidence "+eid.String()+" is archived")
			}
			if err := s.approvals.RequireEvidenceUse(ctx, userID, app.ID, eid); err != nil {
				return err
			}
			pool = append(pool, e)
		}
		method := domain.MethodRuleBased
		if in.AI != nil {
			method = domain.MethodAIAssisted
		}
		analysis := runGapAnalysis(j, app.ID, pool, method)
		if err := s.analyses.MarkStaleForJob(ctx, userID, jobID, uuid.Nil); err != nil {
			return err
		}
		if err := s.analyses.Create(ctx, analysis); err != nil {
			return err
		}
		now := time.Now().UTC()
		v, _ := json.Marshal(analysis)
		op = &domain.Operation{
			ID: uuid.New(), UserID: userID, Type: domain.OpJobAnalysis,
			ApplicationID: &app.ID, Status: domain.OpSucceeded,
			Result:    &domain.OperationResult{Kind: string(domain.OpJobAnalysis), Value: v},
			CreatedAt: now, UpdatedAt: now,
		}
		return s.ops.Create(ctx, op)
	})
	if err != nil {
		return nil, err
	}
	return op, nil
}

func runGapAnalysis(j *domain.JobPosting, appID uuid.UUID, pool []*domain.CareerEvidence, method domain.AnalysisMethod) *domain.GapAnalysis {
	matched := []domain.RequirementMatch{}
	missing := []string{}
	for _, req := range j.Requirements {
		ids := matchEvidence(req, pool)
		if len(ids) > 0 {
			matched = append(matched, domain.RequirementMatch{Requirement: req, EvidenceIDs: ids})
		} else {
			missing = append(missing, req)
		}
	}
	preferredMissing := []string{}
	for _, pref := range j.Preferred {
		if len(matchEvidence(pref, pool)) == 0 {
			preferredMissing = append(preferredMissing, pref)
		}
	}
	var fit *int
	if len(j.Requirements) > 0 {
		score := len(matched) * 100 / len(j.Requirements)
		fit = &score
	}
	eids := []uuid.UUID{}
	for _, e := range pool {
		eids = append(eids, e.ID)
	}
	return &domain.GapAnalysis{
		ID: uuid.New(), UserID: j.UserID, ApplicationID: appID, JobID: j.ID,
		JobRevision: j.Revision, EvidenceIDs: eids,
		Matched: matched, Missing: missing, PreferredMissing: preferredMissing,
		Risks: nonNil(j.Risks), FitScore: fit, Method: method,
		CreatedAt: time.Now().UTC(),
	}
}

func matchEvidence(requirement string, pool []*domain.CareerEvidence) []string {
	ids := []string{}
	req := strings.ToLower(requirement)
	words := significantWords(req)
	for _, e := range pool {
		hit := false
		for _, sk := range e.Skills {
			if sk != "" && strings.Contains(req, strings.ToLower(sk)) {
				hit = true
				break
			}
		}
		if !hit {
			src := strings.ToLower(e.SourceText)
			for _, w := range words {
				if strings.Contains(src, w) {
					hit = true
					break
				}
			}
		}
		if hit {
			ids = append(ids, e.ID.String())
		}
	}
	return ids
}

func significantWords(s string) []string {
	out := []string{}
	for _, w := range strings.FieldsFunc(s, func(r rune) bool {
		return !(r >= 'a' && r <= 'z') && !(r >= '0' && r <= '9') && r != '+' && r != '#' && !(r >= '가' && r <= '힣')
	}) {
		if len(w) >= 3 {
			out = append(out, w)
		}
	}
	return out
}

func (s *JobService) ListAnalyses(ctx context.Context, userID, jobID uuid.UUID, page domain.PageRequest) (domain.Page[domain.GapAnalysis], error) {
	if _, err := s.Get(ctx, userID, jobID); err != nil {
		return domain.Page[domain.GapAnalysis]{}, err
	}
	limit := page.EffectiveLimit()
	items, err := s.analyses.ListByJob(ctx, userID, jobID, page, limit+1)
	if err != nil {
		return domain.Page[domain.GapAnalysis]{}, domain.Internal()
	}
	return domain.NewPage(items, limit, func(g domain.GapAnalysis) domain.Cursor {
		return domain.Cursor{CreatedAt: g.CreatedAt, ID: g.ID}
	}), nil
}

File: push-be/app/domain/services/job.py
from __future__ import annotations
import re
from datetime import datetime, timezone
from typing import Optional
from uuid import UUID, uuid4

from app.db import DB, NotFoundError
from app.domain import entities as ent
from app.domain.errors import (
    internal, map_revision_err, not_found, revision_conflict, validation_field,
)
from app.domain.services.ai import AIGate
from app.domain.services.approval import ApprovalService
from app.domain.validators import validate_https_url
from app.infrastructure.store_applications import ApplicationStore
from app.infrastructure.store_evidence import EvidenceStore
from app.infrastructure.store_jobs import GapAnalysisStore, JobStore
from app.infrastructure.store_operations import OperationStore
from app.jsonutil import to_jsonable


def _now() -> datetime:
    return datetime.now(timezone.utc)


_WORD_RE = re.compile(r"[a-z0-9+#가-힣]+")


def significant_words(s: str) -> list[str]:
    return [w for w in _WORD_RE.findall(s) if len(w) >= 3]


def match_evidence(requirement: str, pool: list[ent.CareerEvidence]) -> list[str]:
    ids = []
    req = requirement.lower()
    words = significant_words(req)
    for e in pool:
        hit = any(sk and sk.lower() in req for sk in e.skills)
        if not hit:
            src = e.source_text.lower()
            hit = any(w in src for w in words)
        if hit:
            ids.append(str(e.id))
    return ids


def run_gap_analysis(j: ent.JobPosting, app_id: UUID,
                     pool: list[ent.CareerEvidence], method: str) -> ent.GapAnalysis:
    matched, missing = [], []
    for req in j.requirements:
        ids = match_evidence(req, pool)
        if ids:
            matched.append(ent.RequirementMatch(requirement=req, evidence_ids=ids))
        else:
            missing.append(req)
    preferred_missing = [p for p in j.preferred if not match_evidence(p, pool)]
    fit = len(matched) * 100 // len(j.requirements) if j.requirements else None
    return ent.GapAnalysis(
        id=uuid4(), user_id=j.user_id, application_id=app_id, job_id=j.id,
        job_revision=j.revision, evidence_ids=[e.id for e in pool],
        matched=matched, missing=missing, preferred_missing=preferred_missing,
        risks=j.risks or [], fit_score=fit, method=method, created_at=_now())


def validate_job_input(company: str, title: str, kind: str,
                       source_url: Optional[str], source_text: str) -> None:
    if not 1 <= len(company) <= 200:
        raise validation_field("company", "company must be 1-200 characters")
    if not 1 <= len(title) <= 300:
        raise validation_field("title", "title must be 1-300 characters")
    if not ent.valid_job_source_kind(kind):
        raise validation_field("sourceKind", "must be URL, TEXT or DOM")
    if kind != "TEXT" and not source_url:
        raise validation_field("sourceUrl", "required for URL/DOM sources")
    if source_url:
        try:
            validate_https_url(source_url)
        except ValueError:
            raise validation_field("sourceUrl", "must be an http(s) URL without credentials")
    if not 1 <= len(source_text) <= 200000:
        raise validation_field(
            "sourceText", "collected source text is required (1-200000 characters)")


class JobService:
    def __init__(self, db: DB, approvals: ApprovalService, ai: AIGate):
        self.db = db
        self.jobs = JobStore(db)
        self.analyses = GapAnalysisStore(db)
        self.applications = ApplicationStore(db)
        self.evidence = EvidenceStore(db)
        self.approvals = approvals
        self.ops = OperationStore(db)
        self.ai = ai

    async def list(self, user_id: UUID, archived: Optional[bool], query: str, page):
        try:
            return await self.jobs.list(user_id, archived, query, page)
        except Exception:
            raise internal()

    async def get(self, user_id: UUID, id_: UUID) -> ent.JobPosting:
        try:
            return await self.jobs.get(user_id, id_)
        except NotFoundError:
            raise not_found()
        except Exception:
            raise internal()

    async def create(self, user_id: UUID, company: str, title: str, source_kind: str,
                     source_url: Optional[str], source_text: str,
                     requirements: list[str], preferred: list[str],
                     deadline: Optional[datetime], language: str) -> ent.JobPosting:
        validate_job_input(company, title, source_kind, source_url, source_text)
        now = _now()
        j = ent.JobPosting(
            id=uuid4(), user_id=user_id, revision=1, company=company, title=title,
            source_kind=source_kind, source_url=source_url, source_text=source_text,
            requirements=requirements or [], preferred=preferred or [],
            keywords=[], risks=[], deadline=deadline, language=language,
            created_at=now, updated_at=now)
        try:
            await self.jobs.create(j)
        except Exception:
            raise internal()
        return j

    async def patch(self, user_id: UUID, id_: UUID, expected: int,
                    company: Optional[str], title: Optional[str],
                    requirements, preferred, deadline, clear_deadline: bool) -> ent.JobPosting:
        out = None

        async def work():
            nonlocal out
            try:
                j = await self.jobs.get(user_id, id_)
            except NotFoundError:
                raise not_found()
            if company is not None:
                if not 1 <= len(company) <= 200:
                    raise validation_field("company", "company must be 1-200 characters")
                j.company = company
            if title is not None:
                if not 1 <= len(title) <= 300:
                    raise validation_field("title", "title must be 1-300 characters")
                j.title = title
            if requirements is not None:
                j.requirements = requirements or []
            if preferred is not None:
                j.preferred = preferred or []
            if clear_deadline:
                j.deadline = None
            elif deadline is not None:
                j.deadline = deadline
            j.updated_at = _now()
            try:
                await self.jobs.update(j, expected)
            except Exception as err:
                raise map_revision_err(err)
            j.revision = expected + 1
            out = j

        await self.db.do(work)
        return out

    async def analyze(self, user_id: UUID, job_id: UUID, application_id: UUID,
                      expected: int, evidence_ids: list[UUID],
                      ai: Optional[ent.AiOptions]) -> ent.Operation:
        await self.ai.check(user_id, ai)
        op = None

        async def work():
            nonlocal op
            try:
                j = await self.jobs.get(user_id, job_id)
            except NotFoundError:
                raise not_found()
            if j.revision != expected:
                raise revision_conflict(j.revision)
            try:
                app = await self.applications.get(user_id, application_id)
            except NotFoundError:
                raise not_found()
            if app.job_id != job_id:
                raise validation_field(
                    "applicationId", "application does not belong to this job")
            pool = []
            for eid in evidence_ids:
                try:
                    e = await self.evidence.get(user_id, eid)
                except NotFoundError:
                    raise validation_field(
                        "evidenceIds", "evidence " + str(eid) + " not found")
                if e.archived:
                    raise validation_field(
                        "evidenceIds", "evidence " + str(eid) + " is archived")
                await self.approvals.require_evidence_use(user_id, app.id, eid)
                pool.append(e)
            method = ent.METHOD_AI_ASSISTED if ai else ent.METHOD_RULE_BASED
            analysis = run_gap_analysis(j, app.id, pool, method)
            await self.analyses.mark_stale_for_job(user_id, job_id, UUID(int=0))
            await self.analyses.create(analysis)
            now = _now()
            op = ent.Operation(
                id=uuid4(), user_id=user_id, type=ent.OP_JOB_ANALYSIS,
                application_id=app.id, status=ent.OP_SUCCEEDED,
                result=ent.OperationResult(kind=ent.OP_JOB_ANALYSIS,
                                           value=to_jsonable(analysis)),
                created_at=now, updated_at=now)
            await self.ops.create(op)

        await self.db.do(work)
        return op

    async def list_analyses(self, user_id: UUID, job_id: UUID, page):
        await self.get(user_id, job_id)
        try:
            return await self.analyses.list_by_job(user_id, job_id, page)
        except Exception:
            raise internal()
